src/util/getNextThoughtsWithContextChain.js

```diff
@@ -0,0 +1,113 @@
+  return contextSiblings[currentIndex + 1]
+}
+
+const firstChildOfContext = (value, cursor, thoughtIndex) => {
+  const context = pathToContext(cursor)
+  return nextSiblingContext(value, context, thoughtIndex)
+}
+
+const firstChildOfThought = (thoughtsRanked, showHiddenThoughts) => {
+  const contextMeta = meta(thoughtsRanked)
+  const sortPreference = getSortPreference(contextMeta)
+  const children = (sortPreference === 'Alphabetical' ? getThoughtsSorted : getThoughtsRanked)(thoughtsRanked)
+  const notHidden = child => !isFunction(child.value) && !meta(pathToContext(thoughtsRanked).concat(child.value)).hidden
+  const childrenFiltered = showHiddenThoughts ? children : children.filter(notHidden)
+  return childrenFiltered[0]
+}
+
+/** Gets nextThoughts and their respective contextChain */
+export const getNextThoughtsWithContextChain = (value, context, rank, cursor, contextRanked, contextChain) => {
+  const { showHiddenThoughts, thoughtIndex, contextViews } = store.getState()
+  if (!contextChain) {
+    contextChain = splitChain(cursor, contextViews)
+  }
+
+  if (context.length === 0 || cursor.length === 0) {
+    return null
+  }
+
+  // if the cursor is at a thought with context view open, move it into context view - jump in
+  if (isContextViewActive(pathToContext(cursor))) {
+    const withinContextView = isContextViewActive(contextRanked)
+    const thoughtsRanked = withinContextView ? getContextSiblings(context, thoughtIndex) : getThoughtsRanked(context)
+
+    const currentThought = withinContextView ? thoughtsRanked.find(thoughts => thoughts.find(thought => thought.value === value))[0] : thoughtsRanked.find(thought => thought.value === value && thought.rank === rank)
+    const firstChild = firstChildOfContext(value, cursor, thoughtIndex)
+    if (firstChild) {
+      return { nextThoughts: [...firstChild, currentThought], contextChain }
+    }
+    // jump out if there are no context children
+    return getNextThoughtsWithContextChain(value, context, rank, cursor, contextOf(contextRanked), contextOf(contextChain))
+  }
+  // if cursor is within a context view, return the next context
+  else if (isContextViewActive(contextRanked)) {
+    const nextSibling = nextSiblingContext(value, context, thoughtIndex, showHiddenThoughts)
+    const contextRankedHead = head(contextRanked)
+
+    if (nextSibling) {
+      return { nextThoughts: [...nextSibling, contextRankedHead], contextChain: contextOf(contextChain) }
+    }
+    // jump out
+    return getNextThoughtsWithContextChain(contextRankedHead.value, contextOf(context), contextRankedHead.rank, cursor, contextOf(contextRanked), contextOf(contextChain))
+  }
+
+  const thoughtsRanked = cursor || RANKED_ROOT
+  const firstChild = firstChildOfThought(thoughtsRanked, showHiddenThoughts)
+  if (firstChild) {
+    return { nextThoughts: unroot(thoughtsRanked.concat(firstChild)), contextChain: [] }
```

Comment:
Having `return` statements interspersed throughout the function makes it harder to see the control flow. Instead, build up the correct result and return it at the end.